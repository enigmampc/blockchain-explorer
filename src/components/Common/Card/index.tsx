import React, { HTMLAttributes } from 'react'
import styled from 'styled-components'

type CardStyledProps = {
  noPadding: boolean
}

type TitleProps = {
  titleAlign: string
}

interface CardProps extends HTMLAttributes<HTMLDivElement> {
  title?: string
  children: React.ReactNode
  noPadding: boolean
  titleAlign?: string
}

const CardStyled = styled.div<CardStyledProps>`
  background-color: ${props => props.theme.cards.backgroundColor};
  border-radius: ${props => props.theme.cards.borderRadius};
  border: ${props => props.theme.cards.border};
  box-shadow: ${props => props.theme.cards.boxShadow};
  display: flex;
  flex-direction: column;
  ${props =>
    props.noPadding
      ? 'padding: 0'
      : 'padding: ' + props.theme.cards.paddingVertical + ' ' + props.theme.cards.paddingHorizontal + ';'}
`

const Title = styled.h2<TitleProps>`
  color: ${props => props.theme.cards.titleColor};
  font-size: 16px;
  font-weight: 700;
  line-height: 1.31;
  margin: 0;
  text-align: ${props => props.titleAlign};
`

const Body = styled.div`
  color: ${props => props.theme.cards.textColor};
  display: flex;
  flex-direction: column;
  flex-grow: 1;
`

const Card = ({ title, children, noPadding, titleAlign = 'left', ...restProps }: CardProps) => {
  return (
    <CardStyled noPadding={noPadding} {...restProps}>
      {title ? <Title titleAlign={titleAlign}>{title}</Title> : null}
      <Body>{children}</Body>
    </CardStyled>
  )
}

export default Card
