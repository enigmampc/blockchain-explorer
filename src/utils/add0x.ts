const add0x = (hex: string): string => (hex ? (hex.indexOf('0x') === 0 ? hex : `0x${hex}`) : '')

export default add0x
