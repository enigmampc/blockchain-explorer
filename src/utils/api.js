import { formatNumber } from './numeralFormats'

export const cryptoCompare = async () => {
  const url = new URL('https://min-api.cryptocompare.com/data/pricemultifull')
  const params = {
    fsyms: 'ENG',
    tsyms: 'USD',
  }
  url.search = new URLSearchParams(params).toString()
  const {
    RAW: {
      ENG: {
        USD: { PRICE, MKTCAP },
      },
    },
  } = await (await fetch(url)).json().catch(console.error)

  return {
    price: formatNumber(PRICE, 4),
    marketCap: formatNumber(MKTCAP, 0),
  }
}

export const coinMarketCap = async () => {
  const url = new URL('https://pro-api.coinmarketcap.com/v1/cryptocurrency/quotes/latest')
  const params = {
    id: 2044,
  }
  url.search = new URLSearchParams(params).toString()
  const headers = new Headers({
    'X-CMC_PRO_API_KEY': process.env.REACT_APP_COINMARKET_CAP_API_KEY,
  })
  const init = {
    method: 'GET',
    headers,
    cache: 'default',
  }
  const request = new Request(url, init)
  const response = await (await fetch(request)).json()

  if (response.error_code) throw new Error(response.error_message)

  const {
    data: {
      [params.id]: {
        quote: {
          USD: { market_cap, price },
        },
      },
    },
  } = response

  return {
    price: formatNumber(price, 4),
    marketCap: formatNumber(market_cap, 0),
  }
}

export const binance = async () => {
  const binanceURL = 'https://api.binance.com/api/v1/ticker/price?symbol='
  const { price } = await (await fetch(`${binanceURL}ENGBTC`)).json().catch()

  return {
    price: formatNumber(price, 4),
    marketCap: '',
  }
}
