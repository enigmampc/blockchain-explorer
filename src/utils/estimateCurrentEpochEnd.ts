import ethApi from './eth'
import { EpochProps } from '../components/Epochs'

const calculateRangesDurations = (range: number[]) => {
  const acc = []

  for (let i = 0; i < range.length; i += 2) {
    const rangeElement = range[i]
    acc.push(rangeElement - range[i + 1])
  }

  return acc
}

const distanceMedian = (range: number[]) => {
  const durations = calculateRangesDurations(range)
  const length = durations.length

  if (length === 0) return length

  durations.sort((a, b) => a - b)

  const half = Math.floor(length / 2)

  return length % 2 ? durations[half] : Math.floor((durations[half - 1] + durations[half]) / 2)
}

const distanceAverage = (range: number[]) => {
  const durations = calculateRangesDurations(range)
  const length = durations.length

  if (!length) return 0

  const sum = durations.reduce((a, b) => a + b)
  return Math.floor(sum / length)
}

export interface EstimatesCurrentEpochEnd {
  pendingTime: number
  finishBlockNumber: number
}

const estimateCurrentEpochEnd = async (epochs?: EpochProps[], average = false): Promise<EstimatesCurrentEpochEnd> => {
  let pendingTime = 0 // estimated life time for the current epoch
  let finishBlockNumber = 0 // estimated block for when the current epoch may end

  if (!epochs) return { pendingTime, finishBlockNumber }

  const sortedEpochs = epochs.sort((a, b) => +b.id - +a.id)
  const currentEpoch = sortedEpochs.slice(0, 1)[0] // currently active epoch

  // builds an array of the form [10, 1, 21, 11, 40, 22]
  // where 10 is the last block of the first epoch in the collection
  // and 1 is the first block of the first epoch in the collection
  // 21 is the last block of the second epoch and 11 is it's first block and so on
  const blocksRange = sortedEpochs
    .reduce((acc: number[], epoch, index) => {
      if (index !== 0) {
        acc.push(+epoch.startBlockNumber)
        acc.push(+epoch.endBlockNumber)
      }

      return acc
    }, [])
    .reverse()

  try {
    const [currentBlock, epochsRangesTimestamps] = await Promise.all([
      ethApi.getBlockNumber(),
      ethApi.getBatchBlocksTimestamps(blocksRange),
    ])
    const epochBlockCount = average ? distanceAverage(blocksRange) : distanceMedian(blocksRange)
    const epochDuration = average ? distanceAverage(epochsRangesTimestamps) : distanceMedian(epochsRangesTimestamps)
    finishBlockNumber = +currentEpoch.startBlockNumber + epochBlockCount
    const currentBlocksLeft = finishBlockNumber - +currentBlock
    const currentBlocksTimes = epochDuration / epochBlockCount
    pendingTime = Math.floor(currentBlocksLeft * currentBlocksTimes) * 1000 // milliseconds
  } catch (e) {
    console.error('Failed to retrieve blocks tiemstamps', e)
  }

  return { pendingTime, finishBlockNumber }
}

export default estimateCurrentEpochEnd
