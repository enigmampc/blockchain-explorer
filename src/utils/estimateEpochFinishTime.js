import ethApi from './eth'

const calculateRangesDurations = range => {
  const acc = []

  for (let i = 0; i < range.length; i += 2) {
    const rangeElement = range[i]
    acc.push(rangeElement - range[i + 1])
  }

  return acc
}

const distanceMedian = range => {
  const durations = calculateRangesDurations(range)
  const length = durations.length

  if (length === 0) return length

  durations.sort((a, b) => a - b)

  const half = Math.floor(length / 2)

  return length % 2 ? durations[half] : Math.floor((durations[half - 1] + durations[half]) / 2)
}

const distanceAverage = range => {
  const durations = calculateRangesDurations(range)
  const length = durations.length

  if (!length) return 0

  const sum = durations.reduce((a, b) => a + b)
  return Math.floor(sum / length)
}

const estimateEpochFinishTimes = async (epochs = [], average = false) => {
  let finishTime = 0 // when the current epoch may end
  let finishBlock = 0 // where the current epoch may end
  const currentEpoch = epochs.slice(0, 1)[0] // currently active epoch

  // builds an array of the form [10, 1, 21, 11, 22, 40]
  // where 10 is the last block of the first epoch in the collection
  // and 1 is the first block of the first epoch in the collection
  // 21 is the last block of the second epoch and 11 is it's first block and so on
  const blocksRange = epochs
    .reduce((acc, epoch, index) => {
      const current = index === 0
      if (!current) {
        acc.push(+epoch.startBlockNumber) // start block number for current epoch
        acc.push(+epochs[index - 1].startBlockNumber - 1) // finish block number for current epoch
      }
      return acc
    }, [])
    .reverse()

  try {
    const [currentBlock, epochsRangesTimestamps] = await Promise.all([
      ethApi.getBlockNumber(),
      ethApi.getBatchBlocksTimestamps(blocksRange),
    ])

    const epochBlockCount = average ? distanceAverage(blocksRange) : distanceMedian(blocksRange)
    const epochDuration = average ? distanceAverage(epochsRangesTimestamps) : distanceMedian(epochsRangesTimestamps)

    finishBlock = +currentEpoch.startBlockNumber + epochBlockCount
    const currentBlocksLeft = finishBlock - +currentBlock
    const currentBlocksTimes = epochDuration / epochBlockCount

    finishTime = Math.floor(currentBlocksLeft * currentBlocksTimes) * 1000 // milliseconds
  } catch (e) {
    console.error('Failed to retrieve blocks tiemstamps', e)
  }

  return { finishTime, finishBlock }
}

export default estimateEpochFinishTimes
