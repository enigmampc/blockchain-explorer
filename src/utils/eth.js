import Web3 from 'web3'
// import EnigmaContract from './Enigma.json'
import getNetworkDetailsBy from './networks'

export const ENG_DECIMALS = 8

class EthAPI {
  constructor() {
    if (process.env.REACT_APP_ETH_NETWORK_ID) {
      const { url } = getNetworkDetailsBy('id')(process.env.REACT_APP_ETH_NETWORK_ID)
      this.web3 = new Web3(new Web3.providers.HttpProvider(url))
    } else {
      this.web3 = new Web3(new Web3.providers.HttpProvider(process.env.REACT_APP_ETH_URL))
    }
  }

  getBlockNumber() {
    return this.web3.eth.getBlockNumber()
  }

  async getBlockTimestamp(blockNumber) {
    const block = await this.web3.eth.getBlock(blockNumber)
    return (block && block.timestamp) || '0'
  }

  async getBatchBlocksTimestamps(range) {
    const batch = new this.web3.BatchRequest()

    const blocks = range.map(blockNumber => {
      return new Promise((resolve, reject) => {
        const request = this.web3.eth.getBlock.request(blockNumber, (error, data) => {
          if (error) {
            reject(error)
          } else {
            resolve(data)
          }
        })
        batch.add(request)
      })
    })

    batch.execute()

    return (await Promise.all(blocks)).map(({ timestamp }) => timestamp)
  }
}

export default new EthAPI()
