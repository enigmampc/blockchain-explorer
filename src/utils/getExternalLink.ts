import getNetworkDetailsBy from './networks'

const { name: networkName } = getNetworkDetailsBy('id')(process.env.REACT_APP_ETH_NETWORK_ID)

const getExternalLink = (ethContract: string) => {
  const subdomain = ['mainnet', 'local', 'unknown'].includes(networkName) ? '' : `${networkName}.`
  return `https://${subdomain}${process.env.REACT_APP_ETHERSCAN_URL}/${ethContract}`
}

export default getExternalLink
