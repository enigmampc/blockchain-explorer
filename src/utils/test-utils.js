import React from 'react'
import '@testing-library/jest-dom/extend-expect'
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { Router } from 'react-router-dom'
import { createBrowserHistory } from 'history'
import theme from '../theme'

const getStyledComponents = StyledComponent => {
  const componentClass = StyledComponent().type.styledComponentId
  return document.getElementsByClassName(componentClass)
}

const history = createBrowserHistory()

const AllTheProviders = ({ children }) => (
  <ThemeProvider theme={theme}>
    <Router history={history}>{children}</Router>
  </ThemeProvider>
)

const customRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options })

export * from '@testing-library/react'

export { customRender as render, getStyledComponents, history }
