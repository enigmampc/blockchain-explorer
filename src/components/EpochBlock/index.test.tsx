import React from 'react'
import { render } from '../../utils/test-utils'
import Card from '.'

it('renders Card without title', () => {
  render(<Card noPadding={false}>This is the card content</Card>)
})

it('renders Card with title', () => {
  render(
    <Card noPadding={false} title="Card Title">
      This is the card content
    </Card>,
  )
})

it('renders Card with title and 0 padding', () => {
  render(
    <Card noPadding={true} title="Card Title">
      This is the card content
    </Card>,
  )
})
