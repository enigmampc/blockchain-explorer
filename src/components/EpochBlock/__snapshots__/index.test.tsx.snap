// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders current EpochBlock 1`] = `
.c0 {
  background-color: #fff;
  border-radius: 5px;
  border: solid 1px #efefef;
  box-shadow: 0 0 8px 0 rgba(0,0,0,0.04);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 15px 18px;
}

.c2 {
  color: #000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c13 {
  color: #999;
  font-size: 14px;
  font-weight: normal;
  line-height: 1.4;
  margin: 0;
  text-align: center;
  text-transform: uppercase;
}

.c12 {
  color: #333;
  font-size: 20px;
  font-weight: 600;
  line-height: 1.4;
  margin: 0;
  text-align: center;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c14 {
  color: #333;
  font-size: 20px;
  font-weight: 600;
  line-height: 1.4;
  margin: 0;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c15 {
  color: #333;
  font-size: 20px;
  font-weight: 600;
  line-height: 1.4;
  margin: 0;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #333;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c15 > img {
  margin-right: 5px;
}

.c16 {
  color: #999;
  font-size: 14px;
  font-weight: normal;
  line-height: 1.4;
  margin: 0;
  text-align: center;
  text-transform: uppercase;
  color: #999;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #eaeaea;
  border-radius: 50%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 116px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  width: 116px;
}

.c5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #fff;
  border-radius: 50%;
  box-shadow: 0 0 10px 0 rgba(0,0,0,0.25);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  width: 100px;
  z-index: 5;
}

.c6 {
  color: #333;
  font-size: 22px;
  font-weight: 600;
  line-height: 1.36;
  margin: 0;
}

.c7 {
  color: #999;
  font-size: 9px;
  font-weight: 600;
  line-height: 1.44;
  margin: 0;
  text-transform: uppercase;
}

.c8 {
  height: 116px;
  left: 0;
  position: absolute;
  top: 0;
  width: 116px;
  z-index: 1;
}

.c9 {
  -webkit-transform-origin: 50% 50%;
  -ms-transform-origin: 50% 50%;
  transform-origin: 50% 50%;
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
  -webkit-transition: 0.25s stroke-dashoffset;
  transition: 0.25s stroke-dashoffset;
}

.c1 {
  cursor: pointer;
  padding: 19px 15px 12px;
  position: relative;
  -webkit-transition: box-shadow 0.15s linear;
  transition: box-shadow 0.15s linear;
}

.c1::before {
  background-color: #e72e9d;
  border-top-left-radius: 5px;
  border-top-right-radius: 5px;
  content: '';
  height: 6px;
  left: 0;
  position: absolute;
  top: -1px;
  width: 100%;
}

.c1:hover {
  box-shadow: 0 0 8px 0 rgba(0,0,0,0.15);
}

.c1 > div {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c4 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin-right: 15px;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c11 {
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-column-gap: 20px;
  margin-bottom: 10px;
}

.c11 > div {
  position: relative;
}

.c11 > div:first-child:after {
  background-color: #d5d5d5;
  content: '';
  height: 34px;
  position: absolute;
  right: -10px;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  width: 1px;
}

<div>
  <div
    class="c0 c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3 c4"
        height="116"
        width="116"
      >
        <div
          class="c5"
          height="100"
          width="100"
        >
          <h3
            class="c6"
          >
            88
            %
          </h3>
          <p
            class="c7"
          >
            Completed Tasks
          </p>
        </div>
        <svg
          class="c8"
          height="116"
          width="116"
        >
          <circle
            class="c9"
            cx="58"
            cy="58"
            fill="transparent"
            r="50"
            stroke="rgba(231, 46, 157, 0.6)"
            stroke-dasharray="314.1592653589793 314.1592653589793"
            stroke-dashoffset="37.6991118430775"
            stroke-width="16"
          />
        </svg>
      </div>
      <div
        class="c10"
      >
        <div
          class="c11"
        >
          <div
            class=""
          >
            <p
              class="c12"
            >
              #123
            </p>
            <h3
              class="c13"
            >
              Epoch
            </h3>
          </div>
          <div
            class=""
          >
            <p
              class="c14"
            >
              150
            </p>
            <h3
              class="c13"
            >
              Tasks
            </h3>
          </div>
        </div>
        <div
          class=""
        >
          <p
            class="c15"
          >
            <img
              alt=""
              src="time.svg"
            />
            30 min
          </p>
          <h3
            class="c16"
          >
            Time Left
          </h3>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders ended EpochBlock 1`] = `
.c0 {
  background-color: #fff;
  border-radius: 5px;
  border: solid 1px #efefef;
  box-shadow: 0 0 8px 0 rgba(0,0,0,0.04);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 15px 18px;
}

.c2 {
  color: #000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c13 {
  color: #999;
  font-size: 14px;
  font-weight: normal;
  line-height: 1.4;
  margin: 0;
  text-align: center;
  text-transform: uppercase;
}

.c12 {
  color: #333;
  font-size: 20px;
  font-weight: 600;
  line-height: 1.4;
  margin: 0;
  text-align: center;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c14 {
  color: #333;
  font-size: 20px;
  font-weight: 600;
  line-height: 1.4;
  margin: 0;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c15 {
  color: #333;
  font-size: 20px;
  font-weight: 600;
  line-height: 1.4;
  margin: 0;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #999;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c15 > img {
  margin-right: 5px;
}

.c16 {
  color: #999;
  font-size: 14px;
  font-weight: normal;
  line-height: 1.4;
  margin: 0;
  text-align: center;
  text-transform: uppercase;
  color: #999;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #eaeaea;
  border-radius: 50%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 116px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  width: 116px;
}

.c5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #fff;
  border-radius: 50%;
  box-shadow: 0 0 10px 0 rgba(0,0,0,0.25);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  width: 100px;
  z-index: 5;
}

.c6 {
  color: #333;
  font-size: 22px;
  font-weight: 600;
  line-height: 1.36;
  margin: 0;
}

.c7 {
  color: #999;
  font-size: 9px;
  font-weight: 600;
  line-height: 1.44;
  margin: 0;
  text-transform: uppercase;
}

.c8 {
  height: 116px;
  left: 0;
  position: absolute;
  top: 0;
  width: 116px;
  z-index: 1;
}

.c9 {
  -webkit-transform-origin: 50% 50%;
  -ms-transform-origin: 50% 50%;
  transform-origin: 50% 50%;
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
  -webkit-transition: 0.25s stroke-dashoffset;
  transition: 0.25s stroke-dashoffset;
}

.c1 {
  cursor: pointer;
  padding: 19px 15px 12px;
  position: relative;
  -webkit-transition: box-shadow 0.15s linear;
  transition: box-shadow 0.15s linear;
}

.c1::before {
  background-color: rgba(28,168,248,0.5);
  border-top-left-radius: 5px;
  border-top-right-radius: 5px;
  content: '';
  height: 6px;
  left: 0;
  position: absolute;
  top: -1px;
  width: 100%;
}

.c1:hover {
  box-shadow: 0 0 8px 0 rgba(0,0,0,0.15);
}

.c1 > div {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c4 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin-right: 15px;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c11 {
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-column-gap: 20px;
  margin-bottom: 10px;
}

.c11 > div {
  position: relative;
}

.c11 > div:first-child:after {
  background-color: #d5d5d5;
  content: '';
  height: 34px;
  position: absolute;
  right: -10px;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  width: 1px;
}

<div>
  <div
    class="c0 c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3 c4"
        height="116"
        width="116"
      >
        <div
          class="c5"
          height="100"
          width="100"
        >
          <h3
            class="c6"
          >
            88
            %
          </h3>
          <p
            class="c7"
          >
            Completed Tasks
          </p>
        </div>
        <svg
          class="c8"
          height="116"
          width="116"
        >
          <circle
            class="c9"
            cx="58"
            cy="58"
            fill="transparent"
            r="50"
            stroke="rgba(28, 168, 248, 0.5)"
            stroke-dasharray="314.1592653589793 314.1592653589793"
            stroke-dashoffset="37.6991118430775"
            stroke-width="16"
          />
        </svg>
      </div>
      <div
        class="c10"
      >
        <div
          class="c11"
        >
          <div
            class=""
          >
            <p
              class="c12"
            >
              #123
            </p>
            <h3
              class="c13"
            >
              Epoch
            </h3>
          </div>
          <div
            class=""
          >
            <p
              class="c14"
            >
              150
            </p>
            <h3
              class="c13"
            >
              Tasks
            </h3>
          </div>
        </div>
        <div
          class=""
        >
          <p
            class="c15"
          >
            <img
              alt=""
              src="lock.svg"
            />
            30 min
          </p>
          <h3
            class="c16"
          >
            Ended
          </h3>
        </div>
      </div>
    </div>
  </div>
</div>
`;
